xarray/core/merge.py

```diff
@@ -533,9 +533,14 @@ def merge(objects, compat='no_conflicts', join='outer', fill_value=dtypes.NA):
     from .dataarray import DataArray
     from .dataset import Dataset
 
-    dict_like_objects = [
-        obj.to_dataset() if isinstance(obj, DataArray) else obj
-        for obj in objects]
+    dict_like_objects = list()
+    for obj in objects:
+        if not (isinstance(obj, (DataArray, Dataset, dict))):
+            raise ValueError("'objects' must be an iterable containing only "
```

Comment:
Thanks @mathause. Would you mind changing this to `TypeError`? That's generally the appropriate error to raise if input arguments have the wrong type.